File: 0322-coin-change/0322-coin-change.py
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:

        dp = [-1] * (amount + 1)
        dp[0] = 0

        for i in range(1, amount + 1):

            minCoins = float("inf")

            for coin in coins:
               if i - coin >= 0 and dp[i - coin] != -1:
                   minCoins = min(minCoins, 1 + dp[i - coin])
            
            if minCoins != float("inf"):
                dp[i] = minCoins
        
        return dp[-1]




        

File: 0322-coin-change/0322-coin-change.cpp
class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        vector<int> dp(amount + 1, INT_MAX);
        dp[0] = 0;
        for (int a = 1; a < amount + 1; a++) {
            for (auto c: coins) {
                if (a - c >= 0 && dp[a - c] != INT_MAX ) {
                    dp[a] = min(dp[a], 1 + dp[a - c]);
                }
            }
        }
        return dp[amount] == INT_MAX ? -1 : dp[amount];


    }
};